src/extension/wps/wps-core/src/main/java/org/geoserver/wps/gs/GetCoveragesValue.java

```diff
@@ -0,0 +1,230 @@
+        GridCoverage2D gridCoverage2D = null;
+        boolean nodata = false;
+        Number[] evaluationValue = null;
+        try {
+            coordinateReferenceSystem =
+                    crs != null && !crs.isEmpty() ? CRS.decode(crs) : reader2d.getCoordinateReferenceSystem();
+        } catch (FactoryException e) {
+            throw new WPSException("Could not decode CRS " + crs, e);
+        }
+        Position position = new Position2D(coordinateReferenceSystem, x, y);
+        if (!coordinateReferenceSystem.equals(reader2d.getCoordinateReferenceSystem())) {
+            try {
+                MathTransform transform =
+                        CRS.findMathTransform(coordinateReferenceSystem, reader2d.getCoordinateReferenceSystem(), true);
+                transform.transform(position, position);
+            } catch (FactoryException | TransformException e) {
+                throw new WPSException("Unable to convert select coordinate reference system to grid CRS ", e);
+            }
+        }
+        for (String n : names) {
+            ci = catalog.getCoverageByName(n);
+            if (ci == null) {
+                throw new WPSException("Could not find coverage " + n);
+            }
+            reader = ci.getGridCoverageReader(null, null);
+            reader2d = (GridCoverage2DReader) reader;
+            if (reader2d.getOriginalEnvelope().contains(position)) {
+                try {
+                    gridCoverage2D = getGridCoverage2D(reader2d, position);
+                } catch (TransformException e) {
+                    throw new WPSException("Could not transform point " + position, e);
+                }
+                evaluationValue = convertPrimitiveArrayToNumberArray(gridCoverage2D.evaluate(position));
+                nodata = checkNoData(gridCoverage2D, evaluationValue);
+                if (!nodata) {
+                    break;
+                }
+            }
+        }
+        if (gridCoverage2D == null || nodata) {
+            return new ValuesAtPoint(new Double[] {});
+        }
+
+        return new ValuesAtPoint(evaluationValue);
+    }
+
+    private static GridCoverage2D getGridCoverage2D(GridCoverage2DReader reader, Position position)
+            throws IOException, TransformException {
+        GridCoverage2D gridCoverage2D;
+
+        Format format = reader.getFormat();
+        final ParameterValueGroup readParams = format.getReadParameters();
+        final Map<String, Serializable> parameters = CoverageUtils.getParametersKVP(readParams);
+
+        // Calculate the bbox around the point
+        MathTransform worldToGridCorner =
+                reader.getOriginalGridToWorld(PixelInCell.CELL_CORNER).inverse();
+        Coordinate coordinate = new Coordinate(position.getCoordinate()[0], position.getCoordinate()[1]);
+        JTS.transform(coordinate, coordinate, worldToGridCorner);
+        GridEnvelope2D testRange =
```

Comment:
2,2 at the end means having a  2x2 grid (they are width/height).

I think you want to read a 3x3 or a 5x5 (in general, an odd number of pixels), with the center pixel in the position that needs to be sampled.  Otherwise picking the center you'll have a math rounding roulette that will make you land on either the pixel above, or the pixel below (and same for left/right) in a 2x2 grid).

Reply:
done